midend/flattenUnions.cpp

```diff
@@ -0,0 +1,359 @@
+/*
+Copyright 2022 Intel Corp.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "flattenUnions.h"
+
+namespace P4 {
+const IR::MethodCallStatement*
+HandleValidityHeaderUnion::processValidityForStr(const IR::Statement *s, const IR::Member *m,
+                                                 cstring headerElement, cstring setValid) {
+    auto method = new IR::Member(s->srcInfo, new IR::Member(m->expr,IR::ID(headerElement)),
+                                 IR::ID(setValid));
+    auto mc = new IR::MethodCallExpression(s->srcInfo, method,
+                                           new IR::Vector<IR::Argument>());
+    typeMap->setType(method, new IR::Type_Method(IR::Type_Void::get(),
+                     new IR::ParameterList(), setValid));
+    return new IR::MethodCallStatement(mc->srcInfo, mc);
+}
+
+const IR::Node* HandleValidityHeaderUnion::setInvalidforRest(const IR::Statement *s,
+                                           const IR::Member *m, const IR::Type_HeaderUnion *hu,
+                                           cstring exclude, bool setValidforCurrMem) {
+    auto code_block = new IR::IndexedVector<IR::StatOrDecl>;
```

Comment:
How does your backend handle this?
If the validity bits are all in a single word this can be done with a single assignment.
I wonder whether it wouldn't be more efficient to write a back-end specific replacement.